File: .claude/skills/manage-student-error-library/scripts/vision_ocr.m
#import <Foundation/Foundation.h>
#import <Vision/Vision.h>
#import <ImageIO/ImageIO.h>
#import <CoreGraphics/CoreGraphics.h>

static void Fail(NSString *message, int code) {
    fprintf(stderr, "%s\n", message.UTF8String);
    exit(code);
}

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        if (argc != 2) {
            Fail(@"Usage: vision_ocr <image-path>", 2);
        }

        NSString *inputPath = [NSString stringWithUTF8String:argv[1]];
        NSURL *inputURL = [NSURL fileURLWithPath:inputPath];
        CGImageSourceRef source = CGImageSourceCreateWithURL((__bridge CFURLRef)inputURL, NULL);
        if (source == NULL) {
            Fail([NSString stringWithFormat:@"Cannot open image: %@", inputPath], 1);
        }
        CGImageRef image = CGImageSourceCreateImageAtIndex(source, 0, NULL);
        CFRelease(source);
        if (image == NULL) {
            Fail([NSString stringWithFormat:@"Cannot decode image: %@", inputPath], 1);
        }

        VNRecognizeTextRequest *request = [[VNRecognizeTextRequest alloc] init];
        request.recognitionLevel = VNRequestTextRecognitionLevelAccurate;
        request.usesLanguageCorrection = YES;
        request.minimumTextHeight = 0.006;
        request.recognitionLanguages = @[@"zh-Hans", @"zh-Hant", @"en-US"];

        VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCGImage:image options:@{}];
        NSError *visionError = nil;
        BOOL succeeded = [handler performRequests:@[request] error:&visionError];
        CGImageRelease(image);
        if (!succeeded) {
            Fail([NSString stringWithFormat:@"Vision OCR failed: %@", visionError.localizedDescription], 1);
        }

        NSArray<VNRecognizedTextObservation *> *observations = request.results ?: @[];
        observations = [observations sortedArrayUsingComparator:^NSComparisonResult(
            VNRecognizedTextObservation *left,
            VNRecognizedTextObservation *right
        ) {
            CGFloat yGap = fabs(CGRectGetMaxY(left.boundingBox) - CGRectGetMaxY(right.boundingBox));
            if (yGap > 0.012) {
                return CGRectGetMaxY(left.boundingBox) > CGRectGetMaxY(right.boundingBox)
                    ? NSOrderedAscending : NSOrderedDescending;
            }
            return CGRectGetMinX(left.boundingBox) < CGRectGetMinX(right.boundingBox)
                ? NSOrderedAscending : NSOrderedDescending;
        }];

        NSMutableArray *lines = [NSMutableArray array];
        NSMutableArray<NSString *> *textLines = [NSMutableArray array];
        double confidenceSum = 0.0;
        for (VNRecognizedTextObservation *observation in observations) {
            VNRecognizedText *candidate = [observation topCandidates:1].firstObject;
            if (candidate == nil) {
                continue;
            }
            CGRect box = observation.boundingBox;
            [lines addObject:@{
                @"text": candidate.string,
                @"confidence": @(candidate.confidence),
                @"bbox": @{
                    @"x": @(box.origin.x),
                    @"y": @(box.origin.y),
                    @"width": @(box.size.width),
                    @"height": @(box.size.height)
                }
            }];
            [textLines addObject:candidate.string];
            confidenceSum += candidate.confidence;
        }

        NSDictionary *result = @{
            @"schema_version": @1,
            @"engine": @"apple-vision-objc",
            @"input": inputPath,
            @"languages": request.recognitionLanguages ?: @[],
            @"average_confidence": @(lines.count ? confidenceSum / lines.count : 0.0),
            @"text": [textLines componentsJoinedByString:@"\n"],
            @"lines": lines
        };
        NSError *jsonError = nil;
        NSData *data = [NSJSONSerialization dataWithJSONObject:result
                                                       options:NSJSONWritingPrettyPrinted | NSJSONWritingSortedKeys
                                                         error:&jsonError];
        if (data == nil) {
            Fail([NSString stringWithFormat:@"Cannot encode OCR JSON: %@", jsonError.localizedDescription], 1);
        }
        fwrite(data.bytes, 1, data.length, stdout);
        fputc('\n', stdout);
    }
    return 0;
}

File: .claude/skills/manage-student-error-library/scripts/vision_ocr.swift
import Foundation
import Vision
import ImageIO
import CoreGraphics

struct OCRBox: Codable {
    let x: Double
    let y: Double
    let width: Double
    let height: Double
}

struct OCRLine: Codable {
    let text: String
    let confidence: Double
    let bbox: OCRBox
}

struct OCRResult: Codable {
    let schemaVersion: Int
    let engine: String
    let input: String
    let languages: [String]
    let averageConfidence: Double
    let text: String
    let lines: [OCRLine]

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case engine
        case input
        case languages
        case averageConfidence = "average_confidence"
        case text
        case lines
    }
}

func fail(_ message: String, code: Int32 = 1) -> Never {
    FileHandle.standardError.write(Data((message + "\n").utf8))
    exit(code)
}

guard CommandLine.arguments.count == 2 else {
    fail("Usage: swift vision_ocr.swift <image-path>", code: 2)
}

let inputPath = CommandLine.arguments[1]
let inputURL = URL(fileURLWithPath: inputPath)
guard let source = CGImageSourceCreateWithURL(inputURL as CFURL, nil),
      let image = CGImageSourceCreateImageAtIndex(source, 0, nil) else {
    fail("Cannot decode image: \(inputPath)")
}

let request = VNRecognizeTextRequest()
request.recognitionLevel = .accurate
request.usesLanguageCorrection = true
request.minimumTextHeight = 0.006

let preferredLanguages = ["zh-Hans", "zh-Hant", "en-US"]
do {
    let supported = try VNRecognizeTextRequest.supportedRecognitionLanguages(
        for: .accurate,
        revision: request.revision
    )
    let selected = preferredLanguages.filter { supported.contains($0) }
    if !selected.isEmpty {
        request.recognitionLanguages = selected
    }
} catch {
    // Vision can still auto-select a supported language when discovery fails.
}

do {
    let handler = VNImageRequestHandler(cgImage: image, options: [:])
    try handler.perform([request])
} catch {
    fail("Vision OCR failed: \(error.localizedDescription)")
}

let observations = request.results ?? []
let ordered = observations.sorted { left, right in
    let yGap = abs(left.boundingBox.maxY - right.boundingBox.maxY)
    if yGap > 0.012 {
        return left.boundingBox.maxY > right.boundingBox.maxY
    }
    return left.boundingBox.minX < right.boundingBox.minX
}

let lines: [OCRLine] = ordered.compactMap { observation in
    guard let candidate = observation.topCandidates(1).first else { return nil }
    let box = observation.boundingBox
    return OCRLine(
        text: candidate.string,
        confidence: Double(candidate.confidence),
        bbox: OCRBox(
            x: Double(box.origin.x),
            y: Double(box.origin.y),
            width: Double(box.size.width),
            height: Double(box.size.height)
        )
    )
}

let average = lines.isEmpty
    ? 0.0
    : lines.reduce(0.0) { $0 + $1.confidence } / Double(lines.count)
let result = OCRResult(
    schemaVersion: 1,
    engine: "apple-vision",
    input: inputPath,
    languages: request.recognitionLanguages,
    averageConfidence: average,
    text: lines.map(\.text).joined(separator: "\n"),
    lines: lines
)

let encoder = JSONEncoder()
encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
do {
    let data = try encoder.encode(result)
    FileHandle.standardOutput.write(data)
    FileHandle.standardOutput.write(Data("\n".utf8))
} catch {
    fail("Cannot encode OCR result: \(error.localizedDescription)")
}
